File: typescript/src/http-rest.ts
import { serve } from "@hono/node-server"
import { Hono } from "hono"

type User = {
  id: number
  name: string
}

const app = new Hono()
const users = new Map<number, User>()

app.get("/users", ctx => ctx.json([...users.values()]))

app.get("/users/:id", ctx => {
  const user = users.get(Number(ctx.req.param("id")))
  return user ? ctx.json(user) : ctx.notFound()
})

app.post("/users", async ctx => {
  const body = await ctx.req.json()
  const id = users.size + 1
  const user = {
    id,
    name: String(body.name)
  }
  users.set(id, user)
  return ctx.json(user, 201)
})

serve({ fetch: app.fetch, port: 3000 })

File: clojure/src/http_rest.clj
(ns http-rest
  (:require [muuntaja.core :as m]
            [reitit.ring :as ring]
            [reitit.ring.middleware.muuntaja :as muuntaja]
            [reitit.ring.middleware.parameters :as parameters]
            [ring.adapter.jetty :as jetty]))

(def store (atom {}))

(defn- list-users [_]
  {:status 200 :body (vals @store)})

(defn- get-user [{{{:keys [id]} :path} :parameters}]
  (if-let [user (@store (parse-long id))]
    {:status 200 :body user}
    {:status 404 :body nil}))

(defn- create-user [{{:keys [body]} :parameters}]
  (let [user (-> (swap! store
                        (fn [s]
                          (let [id (inc (count s))]
                            (assoc s id {:id id :name (:name body)}))))
                 vals
                 last)]
    {:status 201 :body user}))

(def app
  (ring/ring-handler
    (ring/router
      [["/users"     {:get  list-users
                      :post create-user}]
       ["/users/:id" {:get        get-user
                      :parameters {:path {:id string?}}}]]
      {:data {:muuntaja   m/instance
              :middleware [parameters/parameters-middleware
                           muuntaja/format-middleware]}})))

(defn -main [& _]
  (jetty/run-jetty app {:port 3000 :join? false}))
